### ISSUE TITLE
Tmux becomes highly sluggish after extensive usage, and gobbles 100% of a CPU core

### ISSUE THREAD

**Issue by unphased** [tmux/tmux#2551]
Unfortunately I do not have the usual logging because it takes a lot of usage to get to this point. I do reboot the machine at the end of the day, but what I do is run tmux-resurrect, which restores a significant amount of scrollback history into panes. So what i've been seeing (a number of times by now) is after accumulating 4 windows with about 10 or 15 panes each and having about half of those fill up their 50k line history buffers, which is to say, a lot of content getting pumped through, this starts to happen: 

the tmux server process becomes increasingly sluggish, feeling at first like i'm connected to my server over SSH on wifi and it slowly inching away from the router. Except I'm on the local system, and everything here is gigabit ethernet anyway.

I try things like killing the other windows, but the tmux server cannot be restored back to snappiness without killing it. 

I've tried running `perf top` targeting the process, and I'm seeing definite signs that something is wrong, though it's hard to know with just these samples. The left side is the problematic sluggish server process, without anything actively pushing activity on the terminals within, and yet it continues to peg one core at 100%. The one on the right is looking at a tmux server i started recently which has htop running inside it to give it some activity. 

![image](https://user-images.githubusercontent.com/1542910/106346868-27ef4a80-6288-11eb-9aa5-bb91ef6183cb.png)

I will try to keep the server running if I get any tips on what other troubleshooting steps I could undertake, since I'd like to poke at it. Since I've encountered this behavior quite a few times by now even after rebuilding tmux on master I wonder if I can't help try to fix a bug! 

The sluggish bogged down tmux server also takes a good long while to respond to `tmux ls`, or so it seemed at first. Running it again quickly after sees it return rather quickly. But not always. 

```
29 23:20 ~ ❯❯❯ time tmux -L default ls
0: 1 windows (created Fri Jan 29 13:51:58 2021) (attached)
tmux -L default ls  0.00s user 0.00s system 0% cpu 0.343 total
29 23:20 ~ ❯❯❯ time tmux -L default ls
0: 1 windows (created Fri Jan 29 13:51:58 2021) (attached)
tmux -L default ls  0.00s user 0.00s system 69% cpu 0.003 total
29 23:21 ~ ❯❯❯ time tmux -L default ls
0: 1 windows (created Fri Jan 29 13:51:58 2021) (attached)
tmux -L default ls  0.00s user 0.00s system 0% cpu 0.919 total
```

I found that I was able to run this in a loop and see that the response time is elevated (taking up to 2 seconds to respond to `tmux ls`) whenever the sluggishness happens. 

I also realized that after leaving the problematic server running for a long time, it started to "recover" and was not as sluggish all the time, some of the typing could be very responsive, but it would still have a lot of sluggishness. I find that any scrolling in copy-mode or resizing of panes would reliably make it chug even after shell (and sometimes vim!) panes become temporarily partially snappy again. Also `htop` shows that when this is the case it's not 100%ing CPU, so... that's cool I guess.

I suppose I could learn more about the kernel and learn about how to profile it... And this is where I got a major break in the case, the `strace` of the problematic process!

It has a torrent of calls that look like this: 

```
madvise(0x560db3fde000, 24576, MADV_DONTNEED) = 0
madvise(0x560d9f1d4000, 28672, MADV_DONTNEED) = 0
madvise(0x560da9d39000, 28672, MADV_DONTNEED) = 0
madvise(0x560d99906000, 24576, MADV_DONTNEED) = 0
madvise(0x560db3db5000, 24576, MADV_DONTNEED) = 0
madvise(0x560d9c515000, 28672, MADV_DONTNEED) = 0
madvise(0x560d9c000000, 28672, MADV_DONTNEED) = 0
madvise(0x560d9d897000, 28672, MADV_DONTNEED) = 0
madvise(0x560db4855000, 24576, MADV_DONTNEED) = 0
madvise(0x560dac6d9000, 24576, MADV_DONTNEED) = 0
madvise(0x560db1c1f000, 24576, MADV_DONTNEED) = 0
madvise(0x560d998fe000, 28672, MADV_DONTNEED) = 0
madvise(0x560d9da81000, 24576, MADV_DONTNEED) = 0
madvise(0x560da946c000, 24576, MADV_DONTNEED) = 0
madvise(0x560da58dd000, 28672, MADV_DONTNEED) = 0
madvise(0x560dad005000, 28672, MADV_DONTNEED) = 0
madvise(0x560db5369000, 28672, MADV_DONTNEED) = 0
madvise(0x560d9c1c6000, 28672, MADV_DONTNEED) = 0
madvise(0x560d9d57a000, 28672, MADV_DONTNEED) = 0
madvise(0x560d9d6d7000, 28672, MADV_DONTNEED) = 0
madvise(0x560d9db48000, 24576, MADV_DONTNEED) = 0
madvise(0x560d9dbce000, 28672, MADV_DONTNEED) = 0
madvise(0x560d9eb5b000, 24576, MADV_DONTNEED) = 0
madvise(0x560d9f39d000, 28672, MADV_DONTNEED) = 0
madvise(0x560da0488000, 24576, MADV_DONTNEED) = 0
madvise(0x560da25a6000, 28672, MADV_DONTNEED) = 0
madvise(0x560d9d53d000, 28672, MADV_DONTNEED) = 0
madvise(0x560d9e348000, 28672, MADV_DONTNEED) = 0
madvise(0x560d9e6d9000, 24576, MADV_DONTNEED) = 0
madvise(0x560d9eb69000, 28672, MADV_DONTNEED) = 0
madvise(0x560d9f9d7000, 28672, MADV_DONTNEED) = 0
madvise(0x560d9fb79000, 24576, MADV_DONTNEED) = 0
madvise(0x560db14ac000, 24576, MADV_DONTNEED) = 0
madvise(0x560db627e000, 24576, MADV_DONTNEED) = 0
```

whereas I could not reproduce this from the fresh tmux server. It's very cool to watch the strace respond to keyboard input on a shell running on the new process. 

Whenever strace is actively attached to the problematic tmux server process, this torrent of `madvise` is such that it seems to be hammering I/O so hard that there is no way to interact with it, although it does certainly come back once stopping the `strace`. This is the case even when I combine stderr and redirect `strace` output to a file, which accumulated 25MB and 500k lines of output within about 10 seconds or something like that. I couldn't see it repainting a single thing on my terminal client in that time period. 

Great, so I'm fairly confident that it's possible to, by using tmux heavily, throw it into a state where the madvise call is being called in a tight loop. This does hint that there is indeed a bug somewhere near where madvise is being (caused to be) called.

**Comment by unphased** [comment]
This may be the same underlying problem as https://github.com/tmux/tmux/issues/2306 since I do make rather heavy use of `if-shell` and such. 

Actually, my config is such that I use those so much, and i used to have a shell program that I used to produce some context-sensitive behavior based on things like pane_command and such so I can e.g. hit the same key (my caps lock) to cause vim to exit insert mode and the same key would, when in a shell pane, switch to the next pane. But this sluggishness problem isn't being *triggered* by an if-shell call. The tmux server just ends up doing this on its own by itself.

**Comment by nicm** [comment]
Am I meant to guess the tmux version and what platform this is on? I assume Linux?

tmux never calls madvise itself so this smells like a bug in glibc's malloc implementation which is fairly substandard.

What does strace show immediately before a batch of madvise?

Is this a build after 680e7a382f9ea03568aceaf71f3fbb8ced83f76c? If you remove the call to malloc_trim does the problem go away?

**Comment by nicm** [comment]
If you use pstack on the tmux server while the problem is happening, what does it show? Is it always in the same place or a loop of a few places if you run it multiple times?

**Comment by unphased** [comment]
Yes, it's Ubuntu 18.04 on kernel 4.15, here is an example of the strace in between the `madvise`s: 

```
madvise(0x560db7c31000, 8577024, MADV_DONTNEED) = 0
madvise(0x560daab4c000, 9396224, MADV_DONTNEED) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
getpid()                                = 7840
rt_sigprocmask(SIG_BLOCK, ~[RTMIN RT_1], [], 8) = 0
socketpair(AF_UNIX, SOCK_STREAM, 0, [46, 47]) = 0
clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f5a5b312a10) = 101869
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
close(47)                               = 0
fcntl(46, F_GETFL)                      = 0x2 (flags O_RDWR)
fcntl(46, F_SETFL, O_RDWR|O_NONBLOCK)   = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
getpid()                                = 7840
rt_sigprocmask(SIG_BLOCK, ~[RTMIN RT_1], [], 8) = 0
socketpair(AF_UNIX, SOCK_STREAM, 0, [47, 48]) = 0
clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f5a5b312a10) = 101870
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
close(48)                               = 0
fcntl(47, F_GETFL)                      = 0x2 (flags O_RDWR)
fcntl(47, F_SETFL, O_RDWR|O_NONBLOCK)   = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
ioctl(26, TIOCGPGRP, [83522])           = 0
readlink("/proc/83522/cwd", 0x560d95f51c80, 4096) = -1 EACCES (Permission denied)
ioctl(26, TIOCGSID, [9255])             = 0
readlink("/proc/9255/cwd", "/home/slu/rapidplan3", 4096) = 20
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
madvise(0x560da6df1000, 4096, MADV_DONTNEED) = 0
madvise(0x560daa932000, 4096, MADV_DONTNEED) = 0
```

```
madvise(0x560da36ec000, 4096, MADV_DONTNEED) = 0
madvise(0x560da3a92000, 4096, MADV_DONTNEED) = 0
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=101870, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
write(4, "\21", 1)                      = 1
rt_sigreturn({mask=[]})                 = 0
madvise(0x560da331f000, 4096, MADV_DONTNEED) = 0
madvise(0x560d9d4e5000, 4096, MADV_DONTNEED) = 0
```

And another different one:

```
madvise(0x560db7c31000, 8577024, MADV_DONTNEED) = 0
madvise(0x560daab4c000, 9396224, MADV_DONTNEED) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
ioctl(26, TIOCGPGRP, [83522])           = 0
openat(AT_FDCWD, "/proc/83522/cmdline", O_RDONLY) = 41
fstat(41, {st_mode=S_IFREG|0444, st_size=0, ...}) = 0
read(41, "sudo\0perf\0top\0-p\0007840\0", 1024) = 22
close(41)                               = 0
ioctl(26, TIOCGPGRP, [83522])           = 0
readlink("/proc/83522/cwd", 0x560d95f51c80, 4096) = -1 EACCES (Permission denied)
ioctl(26, TIOCGSID, [9255])             = 0
readlink("/proc/9255/cwd", "/home/slu/rapidplan3", 4096) = 20
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
getpid()                                = 7840
rt_sigprocmask(SIG_BLOCK, ~[RTMIN RT_1], [], 8) = 0
socketpair(AF_UNIX, SOCK_STREAM, 0, [41, 42]) = 0
clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f5a5b312a10) = 101794
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
close(42)                               = 0
fcntl(41, F_GETFL)                      = 0x2 (flags O_RDWR)
fcntl(41, F_SETFL, O_RDWR|O_NONBLOCK)   = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=3545, ...}) = 0
madvise(0x560db3724000, 12288, MADV_DONTNEED) = 0
madvise(0x560db16e2000, 12288, MADV_DONTNEED) = 0
```

I'm fairly certain that this particular build is of commit `5a2db4c7`, which is from Jan 4 2021 and I built it on Jan 5 2021, such is the state of the git repo clone and modified timestamp on my executable.

So yes, it's after `680e7a3`, which is from 4 months ago. I shall revert that commit and build and see if i get the behavior again, thank you.

**Comment by unphased** [comment]
fwiw 4 months feels just right in terms of when i started to observe this behavior.

**Comment by unphased** [comment]
I'm looking into this behavior on my 2nd linux server and pstack doesn't work for me: 

```
8889: tmux
pstack: Input/output error
failed to read target.
```

GDB didn't yield anything useful, every time I got this unhelpful libevent stack:

```
Program received signal SIGINT, Interrupt.
0x00007f64198cd937 in __GI___poll (fds=0x5609edf06af0, nfds=13, timeout=4492) at ../sysdeps/unix/sysv/linux/poll.c:29
29      in ../sysdeps/unix/sysv/linux/poll.c
(gdb) bt
#0  0x00007f64198cd937 in __GI___poll (fds=0x5609edf06af0, nfds=13, timeout=4492) at ../sysdeps/unix/sysv/linux/poll.c:29
#1  0x00007f6419b3f981 in ?? () from /lib/x86_64-linux-gnu/libevent-2.1.so.7
#2  0x00007f6419b36625 in event_base_loop () from /lib/x86_64-linux-gnu/libevent-2.1.so.7
#3  0x00005609ed42d9b7 in proc_loop (tp=0x5609edd61830, loopcb=loopcb@entry=0x5609ed43bb90 <server_loop>) at proc.c:215
#4  0x00005609ed43c311 in server_start (client=0x5609edd60a70, flags=flags@entry=402718720, base=base@entry=0x5609edd60780,
    lockfd=lockfd@entry=5, lockfile=0x5609edd61800 "5") at server.c:234
#5  0x00005609ed3ec3b9 in client_connect (flags=<optimized out>, path=0x5609edd55b70 "/tmp/tmux-1000/default", base=0x5609edd60780)
    at client.c:162
#6  client_main (base=0x5609edd60780, argc=0, argv=0x7ffe8a2222c0, flags=<optimized out>, feat=0) at client.c:293
#7  0x00005609ed3e7f25 in main (argc=0, argv=<optimized out>) at tmux.c:493
(gdb) q
```

**Comment by nicm** [comment]
You can send tmux SIGUSR2 to tell it to start logging, you could try that.

Also you could try attaching oprofile or valgrind. We are trying to find out in which functions tmux is spending its CPU time.

**Comment by nicm** [comment]
I suspect this is malloc_trim, I wonder if glibc will do MADVISE_DONTNEED on all areas of the heap that it doesn't need, even if it has done so already. If your heap becomes sufficiently full of holes then that could be a lot of areas. Did you try to reproduce without it?

Maybe we should call malloc_trim much less often, instead of calling it at grid_destroy perhaps we should call it every 30 seconds or so.

**Comment by unphased** [comment]
I got a `strace -c` summary: 

```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 57.39    2.702174           4    542096           madvise
 39.26    1.848643        9578       193           wait4
  2.96    0.139208        1420        98           clone
  0.14    0.006452           5      1267           stat
  0.03    0.001415           4       334           ioctl
  0.03    0.001363          12       106        21 readlink
  0.03    0.001322          13        98           socketpair
  0.03    0.001214           5       218           close
  0.02    0.001073           5       196           fcntl
  0.02    0.001056           5       196           rt_sigprocmask
  0.02    0.000881           4       205           readv
  0.01    0.000636           6        96           write
  0.01    0.000564          11        49           poll
  0.01    0.000491           5        96         1 rt_sigreturn
  0.01    0.000487           4        98           getpid
  0.01    0.000409           4        97           shutdown
  0.01    0.000259           7        33           writev
  0.00    0.000213           4        43        10 read
  0.00    0.000151           6        23           openat
  0.00    0.000077           3        21        21 kill
  0.00    0.000064           2        23           fstat
  0.00    0.000000           0         1           restart_syscall
------ ----------- ----------- --------- --------- ----------------
100.00    4.708152                545587        53 total
```

Currently looking into obtaining kernel stack traces via `perf trace`. OK I will also try to enable logging. Yep logging works. I can easily also control whether the bad behavior is happening: if i disconnect all clients, it won't happen. Reconnecting and interacting will certainly cause it to start up again. It's easy to see the different behaviors while viewing the strace and the tmux log together, I'm making a video.

**Comment by nicm** [comment]
That is a lot of time in wait4, how long was this measured over? You had tmux start 193 child processes?

**Comment by unphased** [comment]
It was measured over less than a minute... 

Here is the video. You may need to wait a few minutes for YT to finish processing the 1080p quality to make it readable. 

https://youtu.be/r76A0BkdEf0

**Comment by nicm** [comment]
If you have no clients, tmux will not update the status line (or the client itself) so no grids created/destroyed and no malloc_trim.

**Comment by nicm** [comment]
OK well 193 processes in a minute is not too bad that's only four or so per second.

**Comment by nicm** [comment]
Anyway it is only 93 clone so only 93 forks, tmux must have to wait a couple of times per child process.

**Comment by nicm** [comment]
Please try without malloc_trim and see if the problem goes away or changes to something different, if it is the problem it is easy to move it.

**Comment by unphased** [comment]
I think that in my configuration there are a ton of calls to processes to assemble content for populating statuslines and things of that nature. 

Even on a fresh tmux server, I monitored the strace of it and started to see (small in length) cascades of madvise upon starting the first `htop` in that tmux. So I think it's some sort of linear or quadratic accumulation of whatever is causing additional redundant/unnecessary calls to it.

**Comment by unphased** [comment]
Based on my video, the output in the tmux server log that corresponds to the madvise is `utf8_to_data` and `utf8_from_data`... No idea if red herring.

**Comment by unphased** [comment]
Here's what I did, I had it attached and in a state where if i interact in vim in the tmux, I can watch the server log show the pattern that comes up when madvise is cascading. If I attach strace to the tmux server, it is hugely slower... So anyway I just put gdb in a loop: `while true; do sudo gdb -batch -ex bt -p 8889; done` and I combed through the gdb output and found this:

```
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
0x00007f64198d3a4b in madvise () at ../sysdeps/unix/syscall-template.S:78
78      ../sysdeps/unix/syscall-template.S: No such file or directory.
#0  0x00007f64198d3a4b in madvise () at ../sysdeps/unix/syscall-template.S:78
#1  0x00007f6419857156 in mtrim (pad=0, av=0x7f64199a3b80 <main_arena>) at malloc.c:4793
#2  __malloc_trim (s=s@entry=0) at malloc.c:4822
#3  0x00005609ed415dc4 in grid_free_lines (gd=gd@entry=0x5609eeae3370, py=py@entry=0, ny=<optimized out>) at grid.c:269
#4  0x00005609ed4162f5 in grid_destroy (gd=0x5609eeae3370) at grid.c:304
#5  0x00005609ed434843 in screen_free (s=s@entry=0x7ffe8a221900) at screen.c:133
#6  0x00005609ed413857 in format_draw (octx=octx@entry=0x7ffe8a221a20, base=base@entry=0x7ffe8a221a70, available=available@entry=67, expanded=expanded@entry=0x5609f3768aa0 "#[align=left range=left default]#[push-default]#[reverse]#[fg=#606060] 1 #[fg=#303030]#[bg=#606060] 0 #[pop-default]#[norange default]#[list=on align=left]#[list=left-marker]<#[list=right-marker]>#"..., srs=srs@entry=0x5609f2606b80) at format-draw.c:832
#7  0x00005609ed4422fd in status_redraw (c=<optimized out>) at status.c:408
#8  0x00005609ed42fb2d in screen_redraw_update (flags=403013777, c=0x5609f26063e0) at screen-redraw.c:538
#9  screen_redraw_screen (c=c@entry=0x5609f26063e0) at screen-redraw.c:599
#10 0x00005609ed43a86c in server_client_check_redraw (c=0x5609f26063e0) at server-client.c:1940
#11 server_client_loop () at server-client.c:1358
#12 0x00005609ed43bbec in server_loop () at server.c:257
#13 0x00005609ed42d9da in proc_loop (tp=0x5609edd61830, loopcb=loopcb@entry=0x5609ed43bb90 <server_loop>) at proc.c:217
#14 0x00005609ed43c311 in server_start (client=0x5609edd60a70, flags=flags@entry=402718720, base=base@entry=0x5609edd60780, lockfd=lockfd@entry=5, lockfile=0x5609edd61800 "\200\276\062\357\tV") at server.c:234
#15 0x00005609ed3ec3b9 in client_connect (flags=<optimized out>, path=0x5609edd55b70 "/tmp/tmux-1000/default", base=0x5609edd60780) at client.c:162
#16 client_main (base=0x5609edd60780, argc=0, argv=0x7ffe8a2222c0, flags=<optimized out>, feat=0) at client.c:293
#17 0x00005609ed3e7f25 in main (argc=0, argv=<optimized out>) at tmux.c:493
[Inferior 1 (process 8889) detached]
```

**Comment by unphased** [comment]
`malloc_trim` confirmed...

**Comment by nicm** [comment]
I think we should move it to something that runs far less often, but I don't want to add a dedicated timer for it so will need to look and see what is already there than makes sense.

**Comment by unphased** [comment]
Is [this](http://notes.secretsauce.net/notes/2016/04/08_glibc-malloc-inefficiency.html) more or less the reason for introducing malloc_trim?
It seems that it should help to ensure it won't be run too often, but the fact that it eventually consistently takes *so long* to run would indicate that it isn't being configured with suitable settings for this application.

**Comment by nicm** [comment]
How long is it taking though? Try putting a log_debug right before and right after?

Isn't the problem just because it is running at least once a second rather than the time it takes?

**Comment by nicm** [comment]
In your strace of roughly 60 seconds, 542096 madvise calls took 4 micros each. If we say there was about 10000 calls per malloc_trim, then that is about 40 millis. I think we could get away with that if we were only doing it say once a minute instead. The only problem is that people would not see tmux's memory drop until some time later.

**Comment by unphased** [comment]
I'll gladly look into it a bit more, it'll be easy to instrument the malloc_trim call for timing.

**Comment by nicm** [comment]
The reason is that glibc is very bad about releasing memory back to the OS and nobody has ever come up with changes to the malloc configuration that seem to help.

**Comment by nicm** [comment]
Heaps inevitably fragment with small allocations and memory can generally only be freed at the end. There are ways to minimise this but it seems as if glibc either does not use them or does not use them effectively.

MADV_DONTNEED can be used to tell the kernel that heap sections are free without having to move areas after them (which is impossible with malloc). It appears that glibc is walking its free list in malloc_trim and calling madvise on each free section. It doesn't seem as if it remembers having already done this, so it does it again every time.

**Comment by unphased** [comment]
So I've looked into a few things. 

- added instrumentation via https://github.com/unphased/tmux/commit/a9fa54357a018c2352a52a610d80c209d2c730e1
- tuned and added a little to my log inspection tooling
- made a video demonstrating the behavior on a tmux server that is moderately fresh and not perceptibly sluggish yet https://youtu.be/RXrSFgYm4Po

Linked above you can review the code for malloc_trim instrumentation. So i think the elapsed time that I dump there is correctly formatted. As I scan through the output I see that the majority of calls to malloc_trim complete within about 20us with outliers in the 100-200 us range. This seems fine. 

Rather immediately it became suspicious that malloc_trim is being run hundreds of time for each discrete interaction in a tmux pane. So the command I'm using to view the tmux server log is 

    tail -F tmux-server-2654442.log | grep malloc_trim | deriv | count-lap 1

Where `deriv` is the script 

```sh
#!/bin/bash

LASTVALUE=
while read -r line; do
        VALUE=${line%% *}
        # VALUE=$(grep -oE "$@" <<< "$line") # you may want to provide -m1 flag as well for ease of use
        if [ -n "$LASTVALUE" ]; then
                # the field that is extracted is to have its derivative computed, and prepended to output.
                DERIV=$(bc <<< "$VALUE - $LASTVALUE")
                echo "$DERIV $line"
        # else
                # We drop the initial line output since it should make downstream parsing easier typically.
                # echo "initial $line"
        fi
        LASTVALUE="$VALUE"
done
```

and `count-lap` is 

```sh
#!/bin/bash

# This is a log view helper intended for use with `deriv` which automatically performs lap
# counting/measurement. Use me when a log is filtered but still pumping heavily, and this (when
# provided with a trigger specification) keeps track of when batches of output begin and counts
# them and diffs what are usually timestamps against the one from the initial log.

# Give on stdin a stream of input, the first numeric field within is interpreted as deriv's output.
# $1 is a provided lap threshold.

# The output of this program is the input prepended with two more columns: the count of entries in
# this lap, and time elapsed (assuming 2nd field in input is a timestamp) since the first activity
# in this lap.

THRESHOLD="$1"

while read -r line; do
  DERIV=${line%% *}
  REST="${line#* }"
  if (( $(echo "$DERIV > $THRESHOLD" | bc) )) || [ -z "$COUNTER" ]; then
    COUNTER=1
    VALUE_START=${REST%% *}
    echo "new measurement $line"
  else
    VALUE=${REST%% *}
    ELAPSED=$(bc <<< "$VALUE - $VALUE_START")
    echo "$COUNTER $ELAPSED $line"
  fi
  ((COUNTER++))
done
```

The output seen, then, are 3 columns that are prepended to the input, which is the server log filtered to the log line I added in my commit. These 3 columns are based off of simple calculations in seconds from the reported second timestamps from the first column of the tmux server log. The first column is the count of repeated log lines in the current "burst", the second column is the total elapsed time from the first entry in the burst to this entry, and the third column is the delta in time between this log line and the line before it. The delta is used by the count-lap utility to determine based on the threshold provided (here 1 second) when a new "burst" is considered to have started. Note sometimes the tail end of a burst is shown as the new burst starts, that's just due to system log buffering as I'm tailing it. Since the timings computed by my scripts here are based on the printed timestamp rather than actual time of receiving that input, the calculations are legitimate... 

So since i'm not really seeing any monotonically increasing behavior in terms of burst size, and though I reckon I'll have trouble in the future when it becomes more sluggish, to be able to meter the bursts properly, I think I will still try to leave this build running and I'll be able to call this up and get a sense of whether it is the malloc_trim call progressively taking much longer to return from, or that our number of calls out to malloc_trim in a given unit of interaction increases, or both, or neither and between strace and the log it will allow me to inspect what the pattern of activity is when it happens.

**Comment by unphased** [comment]
The malloc_trim calls always come in batches of 7 and sometimes 8 that all run within usually 100us. 

And with a threshold of 0.01s we can see as many as nearly 100 calls that happen without an 0.01s time pause between the calls. 

```
new measurement .011431 1612139328.123750 malloc_trim 0.000056484
2 .000025 .000025 1612139328.123775 malloc_trim 0.000014704
3 .000044 .000019 1612139328.123794 malloc_trim 0.000012533
4 .000064 .000020 1612139328.123814 malloc_trim 0.000012576
5 .000084 .000020 1612139328.123834 malloc_trim 0.000012652
6 .000103 .000019 1612139328.123853 malloc_trim 0.000012254
7 .000122 .000019 1612139328.123872 malloc_trim 0.000012285
8 .001726 .001604 1612139328.125476 malloc_trim 0.000069212
9 .005915 .004189 1612139328.129665 malloc_trim 0.000038605
10 .005937 .000022 1612139328.129687 malloc_trim 0.000013307
11 .005955 .000018 1612139328.129705 malloc_trim 0.000011936
12 .005974 .000019 1612139328.129724 malloc_trim 0.000011862
13 .005993 .000019 1612139328.129743 malloc_trim 0.000012018
14 .006011 .000018 1612139328.129761 malloc_trim 0.000012032
15 .006036 .000025 1612139328.129786 malloc_trim 0.000013766
16 .007599 .001563 1612139328.131349 malloc_trim 0.000058688
17 .011792 .004193 1612139328.135542 malloc_trim 0.000036709
18 .011815 .000023 1612139328.135565 malloc_trim 0.000013518
19 .011834 .000019 1612139328.135584 malloc_trim 0.000012065
20 .011853 .000019 1612139328.135603 malloc_trim 0.000011901
21 .011872 .000019 1612139328.135622 malloc_trim 0.000011938
22 .011890 .000018 1612139328.135640 malloc_trim 0.000011680
23 .011909 .000019 1612139328.135659 malloc_trim 0.000011925
24 .013483 .001574 1612139328.137233 malloc_trim 0.000063003
25 .017701 .004218 1612139328.141451 malloc_trim 0.000053110
26 .017726 .000025 1612139328.141476 malloc_trim 0.000013908
27 .017746 .000020 1612139328.141496 malloc_trim 0.000012852
28 .017766 .000020 1612139328.141516 malloc_trim 0.000012453
29 .017786 .000020 1612139328.141536 malloc_trim 0.000012749
30 .017807 .000021 1612139328.141557 malloc_trim 0.000012541
31 .017829 .000022 1612139328.141579 malloc_trim 0.000013157
32 .019444 .001615 1612139328.143194 malloc_trim 0.000073841
33 .023036 .003592 1612139328.146786 malloc_trim 0.000036695
34 .023055 .000019 1612139328.146805 malloc_trim 0.000011192
35 .023072 .000017 1612139328.146822 malloc_trim 0.000009952
36 .023089 .000017 1612139328.146839 malloc_trim 0.000009640
37 .023106 .000017 1612139328.146856 malloc_trim 0.000009819
38 .023122 .000016 1612139328.146872 malloc_trim 0.000010042
39 .023143 .000021 1612139328.146893 malloc_trim 0.000013583
40 .024292 .001149 1612139328.148042 malloc_trim 0.000036430
41 .027266 .002974 1612139328.151016 malloc_trim 0.000038162
42 .027288 .000022 1612139328.151038 malloc_trim 0.000012656
43 .027306 .000018 1612139328.151056 malloc_trim 0.000010517
44 .027322 .000016 1612139328.151072 malloc_trim 0.000010321
45 .027339 .000017 1612139328.151089 malloc_trim 0.000010301
46 .027356 .000017 1612139328.151106 malloc_trim 0.000010859
47 .027376 .000020 1612139328.151126 malloc_trim 0.000012304
48 .028467 .001091 1612139328.152217 malloc_trim 0.000059235
49 .035090 .006623 1612139328.158840 malloc_trim 0.000016036
50 .035105 .000015 1612139328.158855 malloc_trim 0.000006039
51 .035116 .000011 1612139328.158866 malloc_trim 0.000004774
52 .035125 .000009 1612139328.158875 malloc_trim 0.000004786
53 .035137 .000012 1612139328.158887 malloc_trim 0.000004795
54 .035146 .000009 1612139328.158896 malloc_trim 0.000004832
55 .035157 .000011 1612139328.158907 malloc_trim 0.000005891
56 .036211 .001054 1612139328.159961 malloc_trim 0.000050794
57 .038906 .002695 1612139328.162656 malloc_trim 0.000036062
58 .038928 .000022 1612139328.162678 malloc_trim 0.000012318
59 .038948 .000020 1612139328.162698 malloc_trim 0.000011502
60 .038967 .000019 1612139328.162717 malloc_trim 0.000011121
61 .038982 .000015 1612139328.162732 malloc_trim 0.000009730
62 .038997 .000015 1612139328.162747 malloc_trim 0.000009354
63 .039016 .000019 1612139328.162766 malloc_trim 0.000011286
64 .040045 .001029 1612139328.163795 malloc_trim 0.000057218
65 .042802 .002757 1612139328.166552 malloc_trim 0.000014721
66 .042814 .000012 1612139328.166564 malloc_trim 0.000005235
67 .042824 .000010 1612139328.166574 malloc_trim 0.000004945
68 .042835 .000011 1612139328.166585 malloc_trim 0.000004829
69 .042846 .000011 1612139328.166596 malloc_trim 0.000005105
70 .042856 .000010 1612139328.166606 malloc_trim 0.000005053
71 .042868 .000012 1612139328.166618 malloc_trim 0.000006397
72 .043889 .001021 1612139328.167639 malloc_trim 0.000042393
73 .047371 .003482 1612139328.171121 malloc_trim 0.000017144
74 .047385 .000014 1612139328.171135 malloc_trim 0.000005739
75 .047400 .000015 1612139328.171150 malloc_trim 0.000006280
76 .047415 .000015 1612139328.171165 malloc_trim 0.000005883
77 .047429 .000014 1612139328.171179 malloc_trim 0.000005806
78 .047443 .000014 1612139328.171193 malloc_trim 0.000005765
79 .047460 .000017 1612139328.171210 malloc_trim 0.000007756
80 .049081 .001621 1612139328.172831 malloc_trim 0.000047208
81 .053376 .004295 1612139328.177126 malloc_trim 0.000018673
82 .053392 .000016 1612139328.177142 malloc_trim 0.000005846
83 .053407 .000015 1612139328.177157 malloc_trim 0.000006151
84 .053421 .000014 1612139328.177171 malloc_trim 0.000005601
85 .053435 .000014 1612139328.177185 malloc_trim 0.000005546
86 .053449 .000014 1612139328.177199 malloc_trim 0.000005641
87 .053466 .000017 1612139328.177216 malloc_trim 0.000007686
88 .055038 .001572 1612139328.178788 malloc_trim 0.000050986
89 .059252 .004214 1612139328.183002 malloc_trim 0.000017095
90 .059268 .000016 1612139328.183018 malloc_trim 0.000006301
91 .059284 .000016 1612139328.183034 malloc_trim 0.000006854
92 .059298 .000014 1612139328.183048 malloc_trim 0.000006529
93 .059311 .000013 1612139328.183061 malloc_trim 0.000006003
94 .059324 .000013 1612139328.183074 malloc_trim 0.000006086
95 .059341 .000017 1612139328.183091 malloc_trim 0.000007957
96 .060964 .001623 1612139328.184714 malloc_trim 0.000052741
```

**Comment by unphased** [comment]
Maybe tons of grids are destroyed whenever any rendering happens... seems likely. When doing lots of simple operations it is plausible that tmux will need to re-render nearly everything.

OK so then it would appear that even if we just change calling malloc_trim to happen once each time some user interaction happens (rather than once each time a grid is destroyed, which rendering seems to drigger hundreds of times) we'll get like probably a 200x reduction in jank. 

I could get behind your once-every-30-seconds idea. How about once every time 100MB (and configurable) worth of grids are destroyed? Something like that that tracks actual things that touch memory.

**Comment by nicm** [comment]
No it is when it redraws the status line, it creates and destroys a bunch.

**Comment by nicm** [comment]
We want malloc trim when there is no user activity really but at that time tmux will be sleeping.

**Comment by nicm** [comment]
You can try something like this, although I don't really like it because it is more code than the problem really deserves: [x.diff.txt](https://github.com/tmux/tmux/files/5902632/x.diff.txt)

**Comment by unphased** [comment]
Cool, I'll deploy that and see what it says. Thanks

**Comment by unphased** [comment]
> No it is when it redraws the status line, it creates and destroys a bunch.

OK that makes sense because I think the status line is forced to redraw if the process changes the pane title, which those actions i detailed do change.

> We want malloc trim when there is no user activity really but at that time tmux will be sleeping.

Yes it would be ideal to cause it to run when the user is not actively interacting, so that they are less likely to encounter a slowdown. I would be surprised if a single call (which i understand is sufficient to produce a large memory cleanup at least in theory) could end up taking a problematic amount of time based on what I've seen, which is that it's triggered way too often and only gradually slows things down.

**Comment by unphased** [comment]
I haven't had a chance to deploy and test with the diff you sent yet, but my work machine has been running with the malloc_trim call removed. It seems to be running well. After using it for a day I did run a malloc_trim call injection via gdb, it was able to free around 50% of the memory allocated to the process. Clearly it's worth using, and seems like we'd gain most of its benefits even if it were run only once an hour.

**Comment by timotheecour** [comment]
is there any chance this relates to https://github.com/tmux/tmux/issues/2253 ?

**Comment by unphased** [comment]
Hi @timotheecour I can see in that issue there are two behaviors described. neither look anything like this one here. I think it's highly unlikely that they're related. I looked at the large logfile you posted for the format_expand1 issue and I recommend trying to adjust your expansions variously to see what's causing it. the log just seems to indicate it's maybe getting bogged down formatting strings.

**Comment by nicm** [comment]
I think this would be better to run it every hour, but I'd like to think of something else to go in the event because I don't want to add it just for this really: [x.diff.txt](https://github.com/tmux/tmux/files/6020779/x.diff.txt)

**Comment by unphased** [comment]
Thanks @nicm that diff looks very safe for putting in master TBH. I'm running it on my main system now and will definitely report back if I notice anything happening, especially now that I plan to run linux overnight indefinitely again.

**Comment by nicm** [comment]
I have applied this now, thanks!
